text/0000-macro-expansion-for-macro-input.md

```diff
@@ -0,0 +1,214 @@
+
+## Macro calls in macro input
+
+When implementing a procedural or attribute macro you should account for the possibility that a user might provide a macro call in their input. As an example of where this might trip you up when writing a procedural macro, here's a silly one that evaluates to the length of the string literal passed in:
+
+```rust
+extern crate syn;
+#[macro_use]
+extern crate quote;
+
+#[proc_macro]
+pub fn string_length(tokens: TokenStream) -> TokenStream {
+    let lit: syn::LitStr = syn::parse(tokens).unwrap();
+    let len = str_lit.value().len();
+    
+    quote!(#len)
+}
+```
+
+If you call `string_length!` with something obviously wrong, like `string_length!(struct X)`, you'll get a parser error when `unwrap` gets called, which is expected. But what do you think happens if you call `string_length!(stringify!(struct X))`?
+
+It's reasonable to expect that `stringify!(struct X)` gets expanded and turned into a string literal `"struct X"`, before being passed to `string_length`. However, in order to give the most control to proc macro authors, Rust usually doesn't touch any of the ingoing tokens passed to a procedural macro.
+
+A similar issue happens with attribute macros, but in this case there are two places you have to watch out: the attribute arguments, as well as the body. Consider this:
+
+
+```rust
+#[my_attr_macro(value = concat!("Hello, ", "world!"))]
+mod whatever {
+    procedural_macro_that_defines_a_struct! {
+        ...
+    }
+}
+```
+
+If `#[my_attr_macro]` is expecting to see a struct inside of `mod whatever`, it's going to run into trouble when it sees that macro instead. The same happens with `concat!` in the attribute arguments: Rust doesn't look at the input tokens, so it doesn't even know there's a macro to expand!
+
+Thankfully, there's a way to _tell_ Rust to treat some tokens as macros, and to expand them before trying to expand _your_ macro. But first, we need to understand how Rust finds and expands macros.
+
+## Macro expansion and marking
+
+Rust uses an iterative process to expand macros. The loop that Rust performs is roughly as follows:
+
+1. Look for and expand any macros we can parse in expression, item, or attribute position.
+    - Skip any macros that have explicitly marked macros in their arguments.
+    - Are there any new macros we can parse and expand? Go back to step 1.
+2. Look for and expand any explicitly marked macros where there are raw tokens (like inside the arguments to a proc macro or attribute macro).
+    - Are there any new explicitly marked macros we can expand? Go back to step 2.
+    - Otherwise, go back to step 1.
+
+Other than some details about handing macros we can't resolve yet (maybe because they're defined by another macro expansion), that's it!
+
+In order to explicitly mark a macro for the compiler to expand, we actually just mark the `!` or `#` token on the macro call. The compiler looks around the token for the other bits it needs.
```

Comment:
We should also define that the compiler will ignore marked non-macro attributes. This makes it safe to mark attributes, even if a future rustc may decide that, say, `#[boing]` is a valid attribute despite not being expanded by a `proc_macro_attribute` at all. At the very least, it saves us from needing to deny-list `allow`, `warn`, `deny` and `forbid` attributes.

Reply:
Yeah, non-macro attributes are going to be an interesting case. I've added some comments on them.